src/components/tl/mlx5/mcast/tl_mlx5_mcast_allgather.c

```diff
@@ -0,0 +1,439 @@
+    req->proto  = (req->length < comm->max_eager) ? MCAST_PROTO_EAGER :
+                                                    MCAST_PROTO_ZCOPY;
+
+    if (comm->commsize > ONE_SIDED_RELIABILITY_MAX_TEAM_SIZE) {
+        tl_warn(comm->lib,
+                "team size is %d but max supported team size of mcast allgather is %d",
+                comm->commsize, ONE_SIDED_RELIABILITY_MAX_TEAM_SIZE);
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+
+    req->offset      = 0;
+    req->num_packets = (req->length + comm->max_per_packet - 1)/comm->max_per_packet;
+
+    if (req->num_packets == 0) {
+        req->num_packets = 1;
+    }
+
+    ONE_SIDED_MAX_PACKET_COUNT(comm->allgather_comm.max_num_packets);
+
+    if (comm->allgather_comm.max_num_packets < req->num_packets) {
+        tl_warn(comm->lib,
+                "msg size is %ld but max supported msg size of mcast allgather is %d",
+                req->length, comm->allgather_comm.max_num_packets * comm->max_per_packet);
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+
+    req->last_pkt_len = req->length - (req->num_packets - 1)*comm->max_per_packet;
+
+    ucc_assert(req->last_pkt_len > 0 && req->last_pkt_len <= comm->max_per_packet);
+
+    if (req->proto == MCAST_PROTO_ZCOPY) {
+        /* register the send buffer */
+       status = ucc_tl_mlx5_mcast_mem_register(comm->ctx, req->ptr, req->length, &reg);
+       if (UCC_OK != status) {
+            return status;
+       }
+       req->rreg = reg;
+       req->mr   = reg->mr;
+    }
+
+    if (comm->one_sided.reliability_enabled) {
+        req->one_sided_reliability_scheme = (req->length <
+                comm->one_sided.reliability_scheme_msg_threshold) ?
+                ONE_SIDED_ASYNCHRONOUS_PROTO : ONE_SIDED_SYNCHRONOUS_PROTO;
+    } else {
+        req->one_sided_reliability_scheme = ONE_SIDED_NO_RELIABILITY;
+    }
+
+    req->ag_counter = comm->allgather_comm.coll_counter;
+    req->to_send    = req->num_packets;
+    req->to_recv    = comm->commsize * req->num_packets;
+
+    comm->allgather_comm.coll_counter++;
+    return UCC_OK;
+}
+
+static inline ucc_status_t ucc_tl_mlx5_mcast_coll_do_allgather(void* sbuf, void *rbuf, int size,
+                                                               ucc_tl_mlx5_mcast_coll_comm_t *comm,
+                                                               ucc_tl_mlx5_mcast_coll_req_t **task_req_handle)
+{
+    ucc_tl_mlx5_mcast_coll_req_t *req;
+    ucc_status_t                  status;
+
+    tl_trace(comm->lib, "MCAST allgather start, sbuf %p, rbuf %p, size %d, comm %d, "
+             "comm_size %d, counter %d",
+             sbuf, rbuf, size, comm->comm_id, comm->commsize, comm->allgather_comm.coll_counter);
+
+    req = ucc_calloc(1, sizeof(ucc_tl_mlx5_mcast_coll_req_t), "mcast_req");
+    if (!req) {
+        tl_error(comm->lib, "malloc failed");
+        return UCC_ERR_NO_MEMORY;
+    }
+
+    status = ucc_tl_mlx5_mcast_prepare_allgather(sbuf, rbuf, size, comm, req);
```

Comment:
Here also I think it would be clearer having this inlined and not as a separate function

Reply:
sure, i have combined them all into a single function ```ucc_tl_mlx5_mcast_allgather_start```.